### PR TITLE
[GoAth] Google Two-factor verification

### PR DESCRIPTION
**Pull request by kolesnikdi**
for display_qr endpoint make DRF TemplateHTMLRenderer.
For reason:
Just django function() does not want run MIDDLEWARE. So i can't take request.user
But!
Now I can't test this endpoint because TemplateHTMLRenderer works strange

### CODE REVIEW

**Review by ArtemKhriapa — CHANGES_REQUESTED** [review]
Why do you need .log files in the repository?

File: Web_Menu_DA/settings.py [hunk 1]
```diff
@@ -227,3 +225,10 @@
 REDIS_HOST = 'redis_app'
 REDIS_PORT = 6379
 REDIS_DB = 1
+
```
Comment (ArtemKhriapa) [inline Web_Menu_DA/settings.py:228]: cut it

File: two_factor_authentication/type_2fa_mechanics.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+from django.core.cache import cache
+from django.template.loader import render_to_string
+
+from rest_framework import status
+from rest_framework.response import Response
+
+from Web_Menu_DA.settings import CACHE_TIMEOUT_2FA
+from Web_Menu_DA.constants import Types2FA
+
+
```
Comment (ArtemKhriapa) [inline two_factor_authentication/type_2fa_mechanics.py:25]: logger = logging.getLogger(__name__)

logger.error('msg here')

File: two_factor_authentication/type_2fa_mechanics.py
Comment (ArtemKhriapa) [inline two_factor_authentication/type_2fa_mechanics.py:1]: rename to "utils"

File: two_factor_authentication/business_logic.py [hunk 3]
```diff
@@ -1,166 +1,36 @@
...
-
-    return Auth2FAClass.check_2fa(user, received_code)
+import pyotp
+
+from two_factor_authentication.models import GoogleAuth
+from two_factor_authentication.serializers import GoogleAuthSerializer
+from Web_Menu_DA import settings
+from Web_Menu_DA.constants import Types2FA
+
+MSG_TPL = 'Two-factor verification {}'
```
Comment (ArtemKhriapa) [inline two_factor_authentication/business_logic.py:8]: into constants

File: two_factor_authentication/business_logic.py [hunk 4]
```diff
@@ -1,166 +1,36 @@
...
+    elif request_2fa == Types2FA.GAUTH:
+        if previous_gauth := GoogleAuth.objects.filter(owner_id=user.id).last():
+            previous_gauth.is_active = False
+            previous_gauth.save()
+        otp_base32 = pyotp.random_base32()
+        otp_auth_url = pyotp.totp.TOTP(otp_base32).provisioning_uri(name=user.email.lower(), issuer_name="Web_Menu_DA")
+        new_gauth = GoogleAuth.objects.create(owner_id=user.id, otp_base32=otp_base32, otp_auth_url=otp_auth_url)
+        response = GoogleAuthSerializer(instance=new_gauth).data
+        response['msg'] = MSG_TPL.format('with your Google successfully enabled.')
+        response['redirect to'] = f'{settings.HOST}/enable2fa/display_qr'
```
Comment (ArtemKhriapa) [inline two_factor_authentication/business_logic.py:31]: move into serializer

File: two_factor_authentication/serializers.py [hunk 5]
```diff
@@ -1,3 +1,29 @@
...
+                raise serializers.ValidationError({'type_2fa': 'This Choice has been enabled before.'})
+        attrs.pop('password')  # we must delete password for correct work serializer
+        return attrs
+
+
+class GoogleAuthSerializer(serializers.ModelSerializer):
+    owner = serializers.ReadOnlyField(source='owner.email')  # change visible info from owner_id to owner_email
+
+    class Meta:
+        ordering = ['-id']
```
Comment (ArtemKhriapa) [inline two_factor_authentication/serializers.py:27]: delete

File: two_factor_authentication/business_logic.py (hunk 4 above)
Comment (ArtemKhriapa) [inline two_factor_authentication/business_logic.py:31]: reverse

File: two_factor_authentication/business_logic.py [hunk 6]
```diff
@@ -1,166 +1,36 @@
...
+
+from two_factor_authentication.models import GoogleAuth
+from two_factor_authentication.serializers import GoogleAuthSerializer
+from Web_Menu_DA import settings
+from Web_Menu_DA.constants import Types2FA
+
+MSG_TPL = 'Two-factor verification {}'
+
+
+def setup_2fa(request_2fa, user):
```
Comment (ArtemKhriapa) [inline two_factor_authentication/business_logic.py:11]: type2fa

File: Web_Menu_DA/settings.py (hunk 1 above)
Comment (kolesnikdi) [reply Web_Menu_DA/settings.py:228]: done

File: two_factor_authentication/type_2fa_mechanics.py (hunk 2 above)
Comment (kolesnikdi) [reply two_factor_authentication/type_2fa_mechanics.py:25]: done

File: two_factor_authentication/type_2fa_mechanics.py
Comment (kolesnikdi) [reply two_factor_authentication/type_2fa_mechanics.py:1]: move to bl

File: two_factor_authentication/business_logic.py (hunk 3 above)
Comment (kolesnikdi) [reply two_factor_authentication/business_logic.py:8]: moved to two_factor_authentication.constants.py

File: two_factor_authentication/serializers.py (hunk 5 above)
Comment (kolesnikdi) [reply two_factor_authentication/serializers.py:27]: done

File: two_factor_authentication/business_logic.py (hunk 4 above)
Comment (kolesnikdi) [reply two_factor_authentication/business_logic.py:31]: circular import error ----> serializers.URLField(default=reverse('display_qr'))

Comment (kolesnikdi) [reply two_factor_authentication/business_logic.py:31]: circular import error ----> serializers.URLField(default=reverse('display_qr'))

File: two_factor_authentication/business_logic.py (hunk 6 above)
Comment (kolesnikdi) [reply two_factor_authentication/business_logic.py:11]: done

**Review by ArtemKhriapa — APPROVED** [review]

**Pull request merged**